### DISCUSSION TITLE
No Response From SerialServer

### DISCUSSION THREAD

**Discussion by dthomas-sensonix** [pymodbus-dev/pymodbus#1328] [Help] [answered]
I am trying to get the Asynchronous Server example working for a Serial Server.  I have successfully used TCP-based servers in the past to simulate devices, but this time I need to simulate serial devices.  The following client code talks to my physical devices:

```
from pymodbus.client import ModbusSerialClient

client = ModbusSerialClient('/dev/tty.usbserial-FTB75N09', baudrate=19200)
client.connect()
result = client.read_holding_registers(602, 1, 2)
print(result.registers[0])
client.close()
```

I tried to simulate this device by running the `server_async.py` example (with an expanded datastore) with `--comm serial --framer rtu --store sequential --port /dev/tty.usbserial-FTB75N09` as parameters.  Unfortunately I could not get the server to respond to the client.  I also could not find any useful debug information as to why communication was timing out.  

I then tried creating a very simple server to see if that would help:

```
#!/usr/bin/env python3

from pymodbus.datastore import (
    ModbusSequentialDataBlock,
    ModbusServerContext,
    ModbusSlaveContext,
)
from pymodbus.framer.rtu_framer import ModbusRtuFramer
from pymodbus.server import StartSerialServer

store = ModbusSlaveContext(
    di=ModbusSequentialDataBlock(0, [0x0] * 10000),
    co=ModbusSequentialDataBlock(0, [0x0] * 10000),
    hr=ModbusSequentialDataBlock(0, [0x0] * 10000),
    ir=ModbusSequentialDataBlock(0, [0x0] * 10000))
context = ModbusServerContext(slaves=store, single=True)

StartSerialServer(
    context=context,
    port='/dev/tty.usbserial-FTB75N09',
    framer=ModbusRtuFramer,
    baudrate=19200)
```

I observed the client is able to successfully communicate with the devices when the server code is not running.  The client is not able to communicate with the devices when the server code is running, which suggests to me that the server code is connecting to the port.  Unfortunately it does not appear that the server is responding to the client.

Is it possible I have something configured incorrectly?  Should I not use an actual hardware port?  Is there a way I could see what is going wrong?  I understand these scripts are tested as part of the CI testing, but I'm not exactly sure how they are run.  I would gladly replicate those tests on my setup to confirm this isn't a localized problem.  Thanks.

**Comment by janiversen** [comment, marked as answer]
you need 2 ports if you want to test with hardware, one for the server and one for the client.

test_examples.py shows how we test with virtual ports.

**Reply by dthomas-sensonix** [reply]
Thank you.  The plan is not to test with both the Modbus Server and hardware because, as is often the case, there is not enough hardware.  :)

I plan to continue to develop the software surrounding our Modbus Client with actual hardware, but the rest of the team will hopefully be able to interact with the Modbus Server that will simulate the hardware.  I don't think I will need to dynamically update registers, but if I can provide static contents per Server ID then that would really help out.

I'm looking at `test_examples.py` right now.  Is `test_port` virtual?  Do I need to do something to create that (I'm on a Mac)?

**Reply by dthomas-sensonix** [reply]
Okay, I figured out how to set up a virtual port and got it working.  My apologies, I didn't realize that was required, so I attempted to use the physical port for the linkage, but it appears that does not work.  This might complicate deployment, but at least I know how it works now.  Thanks!

**Comment by janiversen** [comment]
Good luck, let us know if you find problems with the library.
